src/game/client/shadow/w3dbuffermanager.cpp

```diff
@@ -0,0 +1,475 @@
+
+        if (ib->m_DX8IndexBuffer == nullptr) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+W3DBufferManager::W3DVertexBufferSlot *W3DBufferManager::Get_Slot(VBM_FVF_TYPES fvf_type, int size)
+{
+    size = (size + 31) & ~31;
+    int size_index = (size / 32) - 1;
+
+    captainslog_dbgassert(size_index < MAX_IB_SIZES && size, "Allocating too large vertex buffer slot");
+
+    W3DVertexBufferSlot *vb_slot = m_W3DVertexBufferSlots[fvf_type][size_index];
+
+    if (vb_slot) {
+        m_W3DVertexBufferSlots[fvf_type][size_index] = vb_slot->m_nextSameSize;
+
+        if (vb_slot->m_nextSameSize) {
+            vb_slot->m_nextSameSize->m_prevSameSize = nullptr;
+        }
+
+        return vb_slot;
+    }
+
+    return Allocate_Slot_Storage(fvf_type, size);
+}
+
+void W3DBufferManager::Release_Slot(W3DVertexBufferSlot *vb_slot)
+{
+    int size_index = (vb_slot->m_size / 32) - 1;
+
+    vb_slot->m_nextSameSize = m_W3DVertexBufferSlots[vb_slot->m_VB->m_format][size_index];
+
+    if (m_W3DVertexBufferSlots[vb_slot->m_VB->m_format][size_index]) {
+        m_W3DVertexBufferSlots[vb_slot->m_VB->m_format][size_index]->m_prevSameSize = vb_slot;
+    }
+
+    m_W3DVertexBufferSlots[vb_slot->m_VB->m_format][size_index] = vb_slot;
+}
+
+W3DBufferManager::W3DVertexBufferSlot *W3DBufferManager::Allocate_Slot_Storage(VBM_FVF_TYPES fvf_type, int size)
+{
+    captainslog_dbgassert(m_numEmptyVertexSlotsAllocated < MAX_NUMBER_SLOTS, "Nore more VB Slots");
+
+    for (W3DVertexBuffer *vb = m_W3DVertexBuffers[fvf_type]; vb; vb = vb->m_nextVB) {
+
+        if (vb->m_size - vb->m_startFreeIndex >= size && m_numEmptyVertexSlotsAllocated < MAX_NUMBER_SLOTS) {
+            W3DVertexBufferSlot *slot = &m_W3DVertexBufferEmptySlots[m_numEmptyVertexSlotsAllocated];
+
+            slot->m_size = size;
+            slot->m_start = vb->m_startFreeIndex;
+            slot->m_VB = vb;
+            slot->m_nextSameVB = vb->m_usedSlots;
+            slot->m_prevSameVB = nullptr;
+
+            if (vb->m_usedSlots) {
+                vb->m_usedSlots->m_prevSameVB = slot;
+            }
+
+            slot->m_nextSameSize = nullptr;
+            slot->m_prevSameSize = nullptr;
+            vb->m_usedSlots = slot;
+            vb->m_startFreeIndex += size;
+            ++m_numEmptyVertexSlotsAllocated;
+
+            return slot;
+        }
+    }
+
+    W3DVertexBuffer *next = m_W3DVertexBuffers[fvf_type];
+
+    captainslog_dbgassert(
+        m_numEmptyVertexBuffersAllocated < MAX_VERTEX_BUFFERS_CREATED, "Reached Max Static VB Shadow Geometry");
+
+    if (m_numEmptyVertexBuffersAllocated < MAX_VERTEX_BUFFERS_CREATED) {
+        m_W3DVertexBuffers[fvf_type] = &m_W3DEmptyVertexBuffers[m_numEmptyVertexBuffersAllocated];
+        m_W3DVertexBuffers[fvf_type]->m_nextVB = next;
+        ++m_numEmptyVertexBuffersAllocated;
+
+        W3DVertexBuffer *buf = m_W3DVertexBuffers[fvf_type];
+
+        int count;
```

Comment:
You can combine these 5 lines in one with `? :` operator. Or use `std::max(8192, size)`.